### ISSUE TITLE
Files stored in non-static paths cannot be deleted.

### ISSUE THREAD

**Issue by ndm2** [FriendsOfCake/cakephp-upload#409]
I've just inherited a project that uses the upload behavior, I have no experience with it yet, so please bear with me in case that might somehow be the intended behavior.

The behavior seems to generate the path of the file to delete dynamically, based on the current configuration, despite the docs saying that in order to be able to delete files at all, the records would need to have a column that holds the path of the file.
- **[/src/Model/Behavior/UploadBehavior.php#L126](https://github.com/josegonzalez/cakephp-upload/blob/3.6.0/src/Model/Behavior/UploadBehavior.php#L126)**
- **[/docs/examples.rst](https://github.com/josegonzalez/cakephp-upload/blob/3.6.0/docs/examples.rst)**

When using a path that contains non-static elements like for example `{year}`, `{month}`, `{day}`, files cannot be deleted at a later point where those values won't be the same anymore as at the point where the record was created.

For example yesterday `webroot{DS}files{DS}{month}{DS}{day}` would have translated to `webroot/files/08/27` where as today it would be `webroot/files/08/28`, and thus deleting a file today that was uploaded yesterday won't work.

**Comment by josegonzalez** [comment]
Good catch, makes a lot of sense. Would you mind contributing a failing pull request for this? I can take a look at it at that point, maybe we can do some regex on the stored path to figure out what the real path should be.

**Comment by ndm2** [comment]
I can give it a try at least...

**Comment by ndm2** [comment]
As a workaround for now I'm using a custom path processor that prefers the entity value over path generation.

```php
public function basepath()
{
	$dirField = Hash::get($this->settings, 'fields.dir', 'dir');

	if ($this->entity->has($dirField) &&
		!empty($this->entity->get($dirField))
	) {
		return $this->entity->get($dirField);
	}

    return parent::basepath();
}
```

That requires some additional security so that user input cannot slip in for the field, but that aside it works OK.

**Comment by ADmad** [comment]
Do we even need to use path processor to get path when deleting file?  Can't one just concat the value for dir field and filename field and pass on that to writer's `delete()`?

**Comment by jorisvaesen** [comment]
@ADmad Was thinking the same, originally made it that way but it got changed in #390 
To me it looks like we just have add to the docs that `keepFilesOnDelete => false` only works when you save the dir(ectory) to the database.

**Comment by ADmad** [comment]
> .. only works when you save the dir(ectory) to the database and always use the dir field when deleting files.

Yeah that is a problem. I personally have modified the plugin to save the path + filename into the filename name if dir field is not used. When using a filesystem abstraction layer like flysystem it becomes imperative to store the full path if you want to delete the file. Also that's why i don't like the config name "dir", it should be "path" instead :)

**Comment by jorisvaesen** [comment]
@ndm2 Could you make a PR with your solution?

**Comment by ndm2** [comment]
@jorisvaesen I could, but I'm not sure that this is really an overly good solution, I've only choosen it because it was one of the few places where I could hook in without having to modify the plugin code.

A better solution would probably be something like an additional method that specifically retrieves the static, stored basepath.

**Comment by ndm2** [comment]
Long time no see, but better late then never :) I'm working on the project again, so I thought I'd give it a try to fix things in the repo.

Please check the updated PR #410

**Comment by ndm2** [comment]
ping @josegonzalez Just want to make sure the update isn't being overlooked, not sure if there's still anyone subscribed to this topic.

**Comment by josegonzalez** [comment]
I'm not working on it at all, but think @ADmad, @jorisvaesen, and @davidyell might have some opinions :)

**Comment by ndm2** [comment]
Ah, sorry, didn't knew that :)

**Comment by davidyell** [comment]
The solutions proposed here all sounds reasonable to me. Storing the path in the entity, and preferring that makes sense to me. It'll have to be persisted somewhere or somehow anyway to be deletable.

**Comment by ADmad** [comment]
Whatever solution @ndm2 comes up with is good for me :)

**Comment by ndm2** [comment]
@davidyell Renaming is probably out of scope of this issue, but I'd agree, `path` is nicer (after all there's path processors, not dir processors ;)).

@ADmad You may want to have a look at #410 then, I guess my comment with the update note got a little bit lost :)

**Comment by jorisvaesen** [comment]
I think it's about time to launch this change, @ndm2 could you fix the travis errors in #410 so it can get merged?

**Comment by ndm2** [comment]
I'm not sure I can, or let's say should... except for the unrelated sniff error, the errors look like they could possibly be related to / fixed by #446
